### PR TITLE
feat(agents): wire HOME path portability + B15/B17/B19/B22 adapter fixes (#70)

### PR DESCRIPTION
**Pull request by chrisleekr**
## Summary

Completes the adapter half of #70: wires `src/core/path-portability` (landed in #72) into every JSON/TOML snapshot+apply path so home-rooted paths round-trip across machines (B24), and ships four independent adapter corrections found in the same bug-bash pass:

- **B15** copilot: agents are single `<name>.agent.md` files per [GitHub Copilot CLI docs](https://docs.github.com/en/copilot/how-tos/copilot-cli/customize-copilot/create-custom-agents-for-cli), not directory tarballs
- **B17** codex: `AGENTS.override.md` is now snapshot and restored alongside `AGENTS.md` (precedence honoured by codex itself on read)
- **B19** claude: `~/.claude/rules/*.md` round-trips so `@~/.claude/rules/*.md` includes in `CLAUDE.md` stay resolvable on the destination machine
- **B22** codex: skills resolve to the canonical `$HOME/.agents/skills` per [Codex Skills docs](https://developers.openai.com/codex/skills); legacy `~/.codex/skills` remains readable on snapshot for forward migration

### `src/core/sanitizer.ts`

The four sanitize helpers (`sanitizeClaudeHooks`, `sanitizeClaudeMcp`, `sanitizeClaudePluginManifest`, `sanitizeClaudePluginMcp`) gain an optional `home: string = homedir()` parameter. They insert `normalizeForVault` between `JSON.parse` and `redactSecretLiterals` so home-rooted string values become the `${AGENTSYNC_HOME}` placeholder before encryption. Tests pass `""` to disable normalization.

### `src/agents/claude.ts`

- Snapshot passes `homedir()` to every `sanitize…` call; marketplace JSON now goes through `normalizeForVault` before redaction
- `applyClaudeHooks` / `applyClaudeMcp` denormalize only the synced subset before merging into the local file, preserving local-only top-level keys with literal `$` values intact
- Plugin manifest, plugin `.mcp.json`, and `marketplace.json` apply now `denormalizeStringFromVault` the raw content before write (whole-file replacements, no subset merge)
- B19: snapshot loop for `rules/*.md` mirroring `commands/*.md`; new `applyClaudeRule` + dispatch in `applyClaudeVault`

### `src/agents/codex.ts`

- `sanitizeCodexConfig` takes `home`, normalizes the parsed TOML before redaction
- `applyCodexConfig` denormalizes incoming TOML pre-merge so the placeholder never reaches disk
- B17: `AGENTS.override.md` snapshot + new `applyCodexAgentsOverrideMd` + dispatch
- B22: snapshot reads both `userSkillsDir` (canonical) and `skillsDir` (legacy); deduplication keys come from the canonical directory listing rather than emitted artifacts so a walker-rejected canonical skill still blocks its legacy copy from leaking through. `applyCodexSkill` writes exclusively to `userSkillsDir`

### `src/agents/cursor.ts`, `src/agents/vscode.ts`

Inline `normalizeForVault` on snapshot (between parse and redact) and `denormalizeStringFromVault` on apply (whole-file write). Cursor's `rules` string flows through `normalizeStringForVault` / `denormalizeStringFromVault`.

### `src/agents/copilot.ts`

B15 rewrite: snapshot iterates `*.agent.md` files individually, emitting plaintext artifacts at `copilot/agents/<name>.agent.md.age`. `applyCopilotAgent` becomes a single-file write. The old per-directory tar archive path, per-agent interior-violation scan, and the `validateSkillName` guard on agent names are removed — secret/never-sync scanning now belongs to the central walker in `commands/push.ts` over the plaintext artifact stream.

### Architecture

**Before — adapter snapshot/apply, no HOME portability:**

```mermaid
flowchart LR
  Disk["~/.claude.json<br/>cwd: /Users/alpha/proj"]:::local --> Parse["JSON.parse"]
  Parse --> Redact["redactSecretLiterals"]:::xform
  Redact --> Encrypt["age encrypt"]:::xform
  Encrypt --> Vault["vault: cwd literal still /Users/alpha/proj"]:::leak
  Vault --> Decrypt["age decrypt"]:::xform
  Decrypt --> Write["atomicWrite as-is"]:::xform
  Write --> Beta["~/.claude.json on machine B<br/>cwd: /Users/alpha/proj BROKEN"]:::leak

  classDef local fill:#ecf0f1,color:#2c3e50,stroke:#2c3e50
  classDef xform fill:#34495e,color:#ffffff,stroke:#2c3e50
  classDef leak fill:#c0392b,color:#ffffff,stroke:#922b21
```

**After — HOME normalize wired between parse and redact; denormalize before write:**

```mermaid
flowchart LR
  Disk2["~/.claude.json<br/>cwd: /Users/alpha/proj"]:::local --> Parse2["JSON.parse"]
  Parse2 --> Norm["normalizeForVault home=/Users/alpha"]:::xform
  Norm --> Redact2["redactSecretLiterals"]:::xform
  Redact2 --> Encrypt2["age encrypt"]:::xform
  Encrypt2 --> Vault2["vault: cwd = AGENTSYNC_HOME/proj"]:::keep
  Vault2 --> Decrypt2["age decrypt"]:::xform
  Decrypt2 --> Denorm["denormalize home=/Users/beta"]:::xform
  Denorm --> Beta2["~/.claude.json on machine B<br/>cwd: /Users/beta/proj"]:::keep

  classDef local fill:#ecf0f1,color:#2c3e50,stroke:#2c3e50
  classDef xform fill:#34495e,color:#ffffff,stroke:#2c3e50
  classDef keep fill:#27ae60,color:#ffffff,stroke:#1e8449
```

## Test Evidence

- `bun run typecheck` — clean
- `bunx biome check src/` — 3 pre-existing warnings on `path-portability.ts` (placeholder constant), no new findings
- `bun test` — **553 pass / 0 fail** (was 535 on main; added 18 new tests covering wiring + B17 + B19 + B22 + dedup-rejection regression + codex TOML round-trip)

New tests:

| File | Coverage |
| --- | --- |
| `src/core/__tests__/sanitizer.test.ts` | 5 tests proving home→placeholder rewrite for each sanitize function |
| `src/agents/__tests__/claude.test.ts` | rules snapshot + apply + vault dispatch; `applyClaudeHooks` / `applyClaudeMcp` denormalize-subset preserves local keys |
| `src/agents/__tests__/codex.test.ts` | `AGENTS.override.md` round-trip; canonical/legacy skills + canonical-wins + canonical-rejected-blocks-legacy regression; config.toml HOME normalize and denormalize end-to-end |
| `src/agents/__tests__/copilot.test.ts` | Replaced 5 tar-layout tests with 2 single-file tests; obsolete interior-violation tests moved to the central-walker contract |

## Risks / Follow-ups

- **B22 migration**: a user with skills only at `~/.codex/skills/<name>/` and `~/.agents/skills/<name>/` non-existent will see the next pull materialize them under `~/.agents/skills/<name>/` (the canonical location). The legacy directory is *not* automatically removed. If the user later re-pushes from a machine where they have both legacy and canonical, the canonical wins.
- **B16 carry-over from #72**: a pre-fix user with `~/.copilot/instructions` (no extension) will see it ignored after this PR — the canonical filename is `copilot-instructions.md`. Documented release-note migration recommended.
- **Markdown out of scope for B24**: `CLAUDE.md`, `AGENTS.md`, rules, commands, agents, prompts are not normalized — prose-style absolute paths inside markdown are user content. The portability invariant applies only to JSON/TOML structured values.
- **B18** (codex `rules/` upstream-source decision) and **e2e harness** (#71) deferred.

## Checklist

- [x] New code has tests where appropriate — 18 new tests covering every wiring point plus the dedup-rejection edge case surfaced in senior review
- [x] Documentation updated if behavior changed — adapter comments inline, WHY-only per project convention

## Summary by CodeRabbit

* **New Features**
  * Claude rules synchronization
  * Codex AGENTS.override.md support and user-scoped skills directory

* **Improvements**
  * Copilot agents now stored as single plaintext agent files
  * Improved HOME-path portability across Claude, Codex, VS Code, Cursor, and Copilot

* **Bug Fixes / Safety**
  * Stricter handling to avoid symlink, traversal, dotfile, and hidden-name leaks during snapshot/restore

[Review Change Stack](https://app.coderabbit.ai/change-stack/chrisleekr/agentsync/pull/73)

### CODE REVIEW

File: src/agents/__tests__/copilot.test.ts [hunk 1]
```diff
@@ -311,18 +262,10 @@ describe("apply* functions", () => {
...
-
-    const buf = await archiveDirectory(srcAgent);
-    await copilotModule.applyCopilotAgent("my-agent", buf.toString("base64"));
+  test("applyCopilotAgent writes a single .agent.md file", async () => {
+    await copilotModule.applyCopilotAgent("my-agent.agent.md", "# Agent content");
 
-    const extracted = await Bun.file(
-      join(testCopilotPaths.agentsDir, "my-agent", "agent.md"),
-    ).text();
+    const extracted = await Bun.file(join(testCopilotPaths.agentsDir, "my-agent.agent.md")).text();
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/agents/__tests__/copilot.test.ts:283]: Fixed in b4e32ce: added regression tests for traversal, dotfile, null-byte, and missing-suffix names against `applyCopilotAgent` (copilot.test.ts).

File: src/agents/claude.ts [hunk 2]
```diff
@@ -107,6 +113,32 @@ export async function snapshotClaude(options: ClaudeSyncOptions = {}): Promise<S
...
+  // includes break on the destination machine.
+  try {
+    const names = await readdir(AgentPaths.claude.rulesDir);
+    for (const name of names) {
+      if (!name.endsWith(".md")) continue;
+      const sourcePath = join(AgentPaths.claude.rulesDir, name);
+      if (shouldNeverSync(sourcePath)) continue;
+      let content: string;
+      try {
+        content = await readFile(sourcePath, "utf8");
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/agents/claude.ts]: Fixed in b4e32ce: switched the rules snapshot to `readdir({withFileTypes:true})` and reject symlinks before `readFile`; added a regression test asserting a symlinked `rules/*.md` produces zero artifacts and its target content never appears in any artifact plaintext.

File: src/agents/copilot.ts [hunk 3]
```diff
@@ -83,43 +77,25 @@ export async function snapshotCopilot(): Promise<SnapshotResult> {
...
+      const sourcePath = join(AgentPaths.copilot.agentsDir, name);
+      if (shouldNeverSync(sourcePath)) continue;
+      const content = await readIfExists(sourcePath);
+      if (content !== null) {
+        artifacts.push({
+          vaultPath: `copilot/agents/${name}.age`,
+          sourcePath,
+          plaintext: content,
+          warnings: [],
+        });
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/agents/copilot.ts]: Fixed in b4e32ce: added `validateCopilotAgentFileName` (rejects path separators, leading dot, null byte, missing `.agent.md` suffix); called from snapshot, `applyCopilotAgent`, and the vault dispatch loop (warn-and-skip on invalid). Snapshot also switched to `readdir({withFileTypes:true})` so symlinked agent files don't smuggle external content. Tests cover all three surfaces.

**Comment by chrisleekr** [comment]
Outside-diff finding from CodeRabbit review #4281971927 addressed in b4e32ce: `collectPluginHooksDir` now calls `normalizeForVault` between parse and redact, and `applyClaudePluginHook` runs `denormalizeStringFromVault` on the raw content before `atomicWrite`. Regression test in claude.test.ts asserts the placeholder round-trips for a plugin hook containing home-rooted `command` and `cwd` values.

**Comment by chrisleekr** [comment]
Outside-diff finding from CodeRabbit review #4282064058 addressed in f0259fd: rewrote the post-B15 adversarial-vault test to use `.hidden.agent.md.age` so it passes the suffix gate and exercises `validateCopilotAgentFileName`. Single-file agents post-B15 have no tar layer, so the payload is plaintext rather than an archive.
